snapcraft/internal/lxd.py

```diff
@@ -40,30 +72,104 @@ def __init__(self, snap_output, tar_filename, project_options):
         self._project_options = project_options
         self._container_name = 'snapcraft-{}'.format(
             petname.Generate(3, '-'))
+        self._find_lxd()
+        self._container = None
+
+    def _find_lxd(self):
+        try:
+            self._client = pylxd.Client()
+        except:
+            try:
+                os.environ['LXD_DIR'] = '/var/snap/lxd/common/lxd'
+                self._client = pylxd.Client()
+            except:
+                raise EnvironmentError(
+                    'The lxd package is not installed, in order to use '
+                    '`cleanbuild` you must install lxd onto your system. '
+                    'Refer to the "Ubuntu Desktop and Ubuntu Server" section '
+                    'on https://linuxcontainers.org/lxd/getting-started-cli/'
+                    '#ubuntu-desktop-and-ubuntu-server to enable a proper '
+                    'setup.')
 
     def _push_file(self, src, dst):
-        check_call(['lxc', 'file', 'push',
-                    src, '{}/{}'.format(self._container_name, dst)])
+        logger.info('Pushing {} to {}'.format(src, dst))
+        with open(src, 'rb') as source_file:
+            data = source_file.read()
+            self._container.files.put(dst, data)
 
     def _pull_file(self, src, dst):
-        check_call(['lxc', 'file', 'pull',
-                    '{}/{}'.format(self._container_name, src), dst])
+        logger.info('Pulling {} from {}'.format(dst, src))
+        with open(dst, 'wb') as destination_file:
+            data = self._container.files.get(src)
+            destination_file.write(data)
 
     def _container_run(self, cmd):
-        check_call(['lxc', 'exec', self._container_name, '--'] + cmd)
+        if not self._container:
+            raise CalledProcessError(-1, cmd)
+        print('Executing {}'.format(cmd))
+        # API call because self._container.execute doesn't expose IO streams
```

Comment:
Can you explain this a little more? `execute()` blocks, so this makes me think that you simply want to be able to show scrolling output while the command executes within the container rather than blocking, but then you also pipe `stdin`. How is that used?